src/Propulsion.SqlStreamStore/StreamReader.fs

```diff
@@ -0,0 +1,226 @@
+
+        member this.UpdateEmptyPage () =
+            pagesEmpty <- pagesEmpty + 1
+            recentPagesEmpty <- recentPagesEmpty + 1
+
+        member this.UpdateCommitedPosition(pos) =
+            lastCommittedPosition <- pos
+
+        member this.UpdateCurMax(cur, max) =
+            currentBatches <- cur
+            maxBatches <- max
+
+        member this.Start =
+            async {
+                let! ct = Async.CancellationToken
+                while not ct.IsCancellationRequested do
+                    report ()
+                    recentPagesRead <- 0
+                    recentPagesEmpty <- 0
+                    do! Async.Sleep statsInterval
+            }
+
+    [<RequireQualifiedAccess>]
+    [<NoComparison>]
+    type Work =
+        | TakeInitial
+        | Page of ReadAllPage
+        | TakeNext of ReadAllPage
+
+    type SubmitBatchHandler =
+        // ingester submit method: epoch * checkpoint * items -> write result
+        int64 * Async<unit> * seq<Propulsion.Streams.StreamEvent<byte[]>> -> Async<int*int>
+
+type StreamReader
+    (
+        logger : ILogger,
+        store: IStreamStore,
+        ledger: ICheckpointer,
+        submitBatch: SubmitBatchHandler,
+        streamId,
+        consumerGroup,
+        ?maxBatchSize: int,
+        ?sleepInterval: TimeSpan,
+        ?statsInterval: TimeSpan,
+        ?stopAfterInterval: TimeSpan
+    ) =
+
+    let maxBatchSize = defaultArg maxBatchSize 100
+    let sleepInterval = defaultArg sleepInterval (TimeSpan.FromSeconds(5.))
+    let stopAfterInterval = defaultArg stopAfterInterval TimeSpan.MaxValue
+
+    let stats = Stats(logger, ?statsInterval = statsInterval)
+
+    let commit batch =
+        async {
+            try
+                do! ledger.CommitPosition { Stream = streamId; ConsumerGroup = consumerGroup; Position = Nullable(batch.lastPosition) }
+                stats.UpdateCommitedPosition(batch.lastPosition)
+                logger.Debug("Committed position {position}", batch.lastPosition)
+            with
+            | exc ->
+                logger.Error(exc, "Exception while commiting position {position}", batch.lastPosition)
+                return! Async.Raise exc
+        }
+
+    let processPage (page: ReadAllPage) =
+        async {
+            if page.Messages.Length > 0 then
+
+                let events =
+                    page.Messages
+                    |> Seq.map StreamMessage.intoStreamEvent
+                    |> Seq.sortBy (fun x -> x.event.Index)
+                    |> Array.ofSeq
+
+                let batch =
+                    {
+                        firstPosition = events.[0].event.Index
+                        lastPosition = events.[events.Length - 1].event.Index
+                        messages = events
+                        isEnd = page.IsEnd
+                    }
+
+                logger.Debug("Submitting a batch of {batchSize} events, position {firstPosition} through {lastPosition}",
```

Comment:
those are not positions - the batch might cross multiple streams

Reply:
In all cases where I'm using the term position, I'm refering to the position in the (non-existent) $all stream.

Reply:
ok, but those values are derived from the `Index` value on the event - those values are the ordinal number of the event within the stream
If that's not the case, we have a different problem, as Propulsion expects `Index` values to be contiguous within a Stream

Reply:
And what happens if they aren't? 

It's a simple change in the event mapping to source it from the stream instead of global position.

Reply:
Position is what's needed for logging in the context of this reader - what you have makes sense overall
But see https://github.com/jet/propulsion/pull/72#discussion_r435193099 for more detail on what purpose/reqs `Index` serves